src/storage.py

```diff
@@ -0,0 +1,267 @@
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Handle Juju Storage Events."""
+
+import itertools
+import json
+import logging
+from subprocess import CalledProcessError, run
+
+from ops.charm import CharmBase, StorageAttachedEvent, StorageDetachingEvent
+from ops.framework import Object, StoredState
+from ops.model import ActiveStatus, MaintenanceStatus
+from ops_sunbeam.guard import BlockedExceptionError, guard
+from tenacity import retry, stop_after_attempt, wait_fixed
+
+import microceph
+
+logger = logging.getLogger(__name__)
+
+
+class StorageHandler(Object):
+    """The Storage class manages the Juju storage events.
+
+    Observes the following events:
+    1) *_storage_attached
+    2) *_storage_detaching
+    """
+
+    name = "storage"
+    osd_data_path = "/var/snap/microceph/common/data/osd"
+
+    charm = None
+    """
+        osd_data: type dict"
+            disk_by_id: OSD disk by id
+            disk: OSD disk path
+            wal: wal disk path
+            db: db disk path
+    """
+    _stored = StoredState()
+
+    def __init__(self, charm: CharmBase, name="storage"):
+        super().__init__(charm, name)
+        self._stored.set_default(osd_data={})
+        self.charm = charm
+        self.name = name
+
+        # Attach handlers
+        self.framework.observe(
+            getattr(charm.on, "osd_devices_storage_attached"), self._on_osd_devices_attached
+        )
+        for key in ["disk", "wal", "db"]:
+            self.framework.observe(getattr(charm.on, f"{key}_storage_attached"), self._on_attached)
+
+        # OSD Detaching handlers.
+        for key in ["osd_devices", "disk", "wal", "db"]:
+            self.framework.observe(
+                getattr(charm.on, f"{key}_storage_detaching"), self._on_storage_detaching
+            )
+
+    """handlers"""
+
+    def _on_attached(self, event: StorageAttachedEvent):
+        """Updates the attached storage devices in state."""
+        if not microceph._is_ready():
+            logger.warning("MicroCeph not ready yet, deferring storage event.")
+            event.defer()
+            return
+
+        self._clean_stale_osd_data()
+
+        enroll = {
+            "disk": [],
+            "wal": [],
+            "db": [],
+        }
+
+        # filter only storage directives for wal/db and disk.
+        accepts = {"disk", "wal", "db"}
+        for storage in [
+            x for x in self.juju_storage_list() if any(x.find(id) >= 0 for id in accepts)
+        ]:
+            # split storage names of the form disk/0
+            directive = "".join(itertools.takewhile(str.isalpha, storage))
```

Comment:
The method `on_attached` is storage attached handler for disk/wal and db devices (but not osd_devices). Since the directives are defined by us and `juju_storage_list` can only provide attached storage names, fetching directive name can be safely performed by iterating until the first non-alpha character. examples: `wal/4, db/7 or disk/1`

Reply:
I've refactored the filtering logic to a separate strict checking func though.